pkg/registry/apis/dashboard/authorizer.go

```diff
@@ -0,0 +1,97 @@
+package dashboard
+
+import (
+	"context"
+
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+
+	"github.com/grafana/authlib/types"
+	"github.com/grafana/grafana/pkg/apimachinery/identity"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/services/dashboards"
+	"github.com/grafana/grafana/pkg/services/guardian"
+	"github.com/grafana/grafana/pkg/storage/legacysql/dualwrite"
+)
+
+func GetAuthorizer(dashboardService dashboards.DashboardService, dualWriter dualwrite.Service, l log.Logger) authorizer.Authorizer {
+	return authorizer.AuthorizerFunc(
+		func(ctx context.Context, attr authorizer.Attributes) (authorized authorizer.Decision, reason string, err error) {
+			// Note that we will return Allow more than expected.
+			// This is because we do NOT want to hit the RoleAuthorizer that would be evaluated afterwards.
+
+			// Check if we're reading from legacy dashboards and folders
+			// This authorizer is only used for mode 0 to 2. Mode 3 onwards, unified storage handles access control.
+			isReadingLegacy := dualwrite.IsReadingLegacyDashboardsAndFolders(ctx, dualWriter)
+			if !isReadingLegacy {
+				return authorizer.DecisionAllow, "relying on unified storage for access control", nil
+			}
+
+			// Use the standard authorizer
+			if !attr.IsResourceRequest() {
+				// TODO: When is this used?
+				return authorizer.DecisionNoOpinion, "", nil
+			}
+
+			user, err := identity.GetRequester(ctx)
+			if err != nil {
+				return authorizer.DecisionDeny, "", err
+			}
+
+			if attr.GetVerb() == "create" {
+				// Permissions will be handled downstream
+				return authorizer.DecisionAllow, "", nil
+			}
```

Comment:
Note: when we move from guardians to using access control service directly, we could check that the user has dashboard create action (without checking the scope) to return early in case they don't have this permission.

Reply:
Good call, will make sure to consider that when I switch to access control!

We can't ever use DecisionNoOpinion because then it will fall back to the raw role authorizer, which is overly restrictive and not aware of the resource-level permissions.